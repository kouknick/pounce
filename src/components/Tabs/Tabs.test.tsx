import React from 'react';
import { render, act, waitFor, AxeResults, fireEvent } from 'test-utils';
import { axe } from 'jest-axe';
import { Tabs, TabList, Tab, TabPanels, TabPanel } from './index';

it('renders', () => {
  const { container } = render(
    <Tabs>
      <TabList>
        <Tab>1</Tab>
        <Tab>2</Tab>
      </TabList>
      <TabPanels>
        <TabPanel>One</TabPanel>
        <TabPanel>Two</TabPanel>
      </TabPanels>
    </Tabs>
  );
  expect(container).toMatchSnapshot();
});

it('functions correctly by default', () => {
  const { getByText } = render(
    <Tabs>
      <TabList>
        <Tab>1</Tab>
        <Tab>2</Tab>
      </TabList>
      <TabPanels>
        <TabPanel>One</TabPanel>
        <TabPanel>Two</TabPanel>
      </TabPanels>
    </Tabs>
  );

  expect(getByText('1')).toHaveAttribute('aria-selected', 'true');
  expect(getByText('One')).not.toHaveAttribute('hidden');
  expect(getByText('2')).toHaveAttribute('aria-selected', 'false');
  expect(getByText('Two')).toHaveAttribute('hidden');

  fireEvent.click(getByText('2'));

  expect(getByText('1')).toHaveAttribute('aria-selected', 'false');
  expect(getByText('One')).toHaveAttribute('hidden');
  expect(getByText('2')).toHaveAttribute('aria-selected', 'true');
  expect(getByText('Two')).not.toHaveAttribute('hidden');
});

it('correctly lazy loads', () => {
  const { getByText, queryByText } = render(
    <Tabs>
      <TabList>
        <Tab>1</Tab>
        <Tab>2</Tab>
      </TabList>
      <TabPanels>
        <TabPanel>One</TabPanel>
        <TabPanel lazy>Two</TabPanel>
      </TabPanels>
    </Tabs>
  );

  expect(getByText('One')).toBeInTheDocument();
  expect(queryByText('Two')).toBeFalsy();

  fireEvent.click(getByText('2'));

  expect(getByText('One')).toBeInTheDocument();
  expect(getByText('Two')).toBeInTheDocument();
});

it('correctly lazy loads and unmounts when inactive', () => {
  const { getByText, queryByText } = render(
    <Tabs>
      <TabList>
        <Tab>1</Tab>
        <Tab>2</Tab>
      </TabList>
      <TabPanels>
        <TabPanel lazy unmountWhenInactive>
          One
        </TabPanel>
        <TabPanel lazy>Two</TabPanel>
      </TabPanels>
    </Tabs>
  );

  expect(getByText('One')).toBeInTheDocument();
  expect(queryByText('Two')).toBeFalsy();

  fireEvent.click(getByText('2'));

  expect(queryByText('One')).toBeFalsy();
  expect(getByText('Two')).toBeInTheDocument();
});

it('aligns with a11y', async () => {
  const { container } = render(
    <Tabs>
      <TabList>
        <Tab>1</Tab>
        <Tab>2</Tab>
      </TabList>
      <TabPanels>
        <TabPanel>One</TabPanel>
        <TabPanel>Two</TabPanel>
      </TabPanels>
    </Tabs>
  );
  let results: AxeResults;
  await act(async () => {
    results = await axe(container);
  });
  await waitFor(() => {
    expect(results).toHaveNoViolations();
  });
});
