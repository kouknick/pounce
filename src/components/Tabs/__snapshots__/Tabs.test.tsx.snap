// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs renders 1`] = `
.pounce-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex-wrap: wrap;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.pounce-0 {
  color: gray-50;
  background-color: transparent;
  z-index: 1;
  cursor: pointer;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-transition: all 0.1s linear;
  transition: all 0.1s linear;
}

.pounce-0:disabled,
.pounce-0:disabled:focus,
.pounce-0:disabled:hover,
.pounce-0[aria-disabled=true],
.pounce-0[aria-disabled=true]:focus,
.pounce-0[aria-disabled=true]:hover {
  opacity: 0.3;
  pointer-events: none;
  cursor: default;
}

.pounce-6 {
  outline: none;
}

<div>
  <div
    data-orientation="horizontal"
    data-reach-tabs=""
  >
    <div
      aria-orientation="horizontal"
      class="pounce-4 pounce-1"
      data-reach-tab-list=""
      role="tablist"
    >
      <button
        aria-controls="tabs--1--panel--0"
        aria-selected="true"
        class="pounce-0 pounce-1"
        data-orientation="horizontal"
        data-reach-tab=""
        data-selected=""
        id="tabs--1--tab--0"
        role="tab"
        tabindex="0"
      >
        1
      </button>
      <button
        aria-controls="tabs--1--panel--1"
        aria-selected="false"
        class="pounce-0 pounce-1"
        data-orientation="horizontal"
        data-reach-tab=""
        id="tabs--1--tab--1"
        role="tab"
        tabindex="-1"
      >
        2
      </button>
    </div>
    <div
      data-reach-tab-panels=""
    >
      <div
        aria-labelledby="tabs--1--tab--0"
        class="pounce-6 pounce-1"
        data-reach-tab-panel=""
        id="tabs--1--panel--0"
        role="tabpanel"
        tabindex="-1"
      >
        One
      </div>
      <div
        aria-labelledby="tabs--1--tab--1"
        class="pounce-6 pounce-1"
        data-reach-tab-panel=""
        hidden=""
        id="tabs--1--panel--1"
        role="tabpanel"
        tabindex="-1"
      >
        Two
      </div>
    </div>
  </div>
</div>
`;
